core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/DedupAggregator.java

```diff
@@ -0,0 +1,187 @@
+    for (String fieldName : uniqueFields) {
+      builder.set(fieldName, record.get(fieldName));
+    }
+    emitter.emit(builder.build());
+  }
+
+  @Override
+  public void aggregate(StructuredRecord groupKey, Iterator<StructuredRecord> iterator,
+                        Emitter<StructuredRecord> emitter) throws Exception {
+    if (!iterator.hasNext()) {
+      return;
+    }
+
+    SelectionFunction selectionFunction;
+    if (filterFunction == null) {
+      emitter.emit(iterator.next());
+    } else {
+      StructuredRecord firstRecord = iterator.next();
+      Schema.Field firstField = firstRecord.getSchema().getField(filterFunction.getField());
+      selectionFunction = filterFunction.getSelectionFunction(firstField.getSchema());
+      selectionFunction.beginFunction();
+      selectionFunction.operateOn(firstRecord);
+
+      while (iterator.hasNext()) {
+        selectionFunction.operateOn(iterator.next());
+      }
+
+      selectionFunction.finishFunction();
+      List<StructuredRecord> outputRecords = selectionFunction.getSelectedRecords();
+      for (StructuredRecord outputRecord : outputRecords) {
+        Schema outputSchema = getOutputSchema(outputRecord.getSchema());
+        validateSchema(outputSchema, uniqueFields, filterFunction);
+        StructuredRecord.Builder builder = StructuredRecord.builder(outputSchema);
+        for (Schema.Field field : outputRecord.getSchema().getFields()) {
+          builder.set(field.getName(), outputRecord.get(field.getName()));
+        }
+        emitter.emit(builder.build());
+      }
+    }
+  }
+
+  @Path("outputSchema")
+  public Schema getOutputSchema(GetSchemaRequest request) {
+    Schema inputSchema;
+    try {
+      inputSchema = Schema.parseJson(request.inputSchema);
+    } catch (Exception e) {
+      throw new BadRequestException("Could not parse input schema " + request.inputSchema);
+    }
+    try {
+      return getOutputSchema(inputSchema);
+    } catch (IllegalArgumentException e) {
+      throw new BadRequestException(e.getMessage(), e);
+    }
+  }
+
+  private Schema getGroupKeySchema(Schema inputSchema) {
+    List<Schema.Field> fields = new ArrayList<>();
+    for (String fieldName : dedupConfig.getUniqueFields()) {
+      Schema.Field field = inputSchema.getField(fieldName);
+      if (field == null) {
+        throw new IllegalArgumentException(String.format("Field %s does not exist in input schema %s.",
+                                                         fieldName, inputSchema));
+      }
+      fields.add(field);
+    }
+    return Schema.recordOf(inputSchema.getRecordName() + ".unique", fields);
+  }
+
+  private Schema getOutputSchema(Schema inputSchema) {
+    return Schema.recordOf(inputSchema.getRecordName() + ".dedup", inputSchema.getFields());
+  }
+
+  /**
+   * Endpoint request for output schema.
+   */
+  public static class GetSchemaRequest extends DedupConfig {
+    private String inputSchema;
```

Comment:
learned recently that this can actually be the 'Schema' object so that the plugin doesn't have to parse stringified json.  Should change the other aggregator to Schema as well